# Analysis of the “Movie IMBD Dataset"

#### Overview :

This project analyzes the “Movie IMBD Dataset” to uncover patterns and insights related to movie ratings, genres, budgets, and social media influence. The work involves data cleaning, preprocessing, exploratory data analysis, and visualization using Python libraries such as Pandas, Matplotlib, Seaborn, and Plotly. By examining correlations, trends, and sentiment indicators, the project aims to highlight factors that contribute to a movie’s success and summarize actionable insights through clear visualizations and interpretations.

### Importing Libraries

In [5]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings 
warnings.filterwarnings("ignore")

In [ ]:
#from IPython.display import FileLink

# Save CSV
#data.to_csv("movie_metadata_duplicate.csv", index=False)

## 1. Data Loading and Initial Overview

##### Loading dataset

In [57]:
data = pd.read_csv("movie_metadata.csv")

##### Describing an overview of Dataset

In [13]:
#first 5 rows
print("=========First 5 rows==========")
data.head(5)

=========First 5 rows==========


color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4    NaN        Doug Walker                     NaN       NaN   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23000.0    Christian Bale   
4                    131.0                     NaN        Rob Walker   

   actor_1_facebook_likes        gross                           genres  ...  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0  309404152.0         Action|Adventure|Fantasy  ...   
2                 11000.0  200074175.0        Action|Adventure|Thriller  ...   
3                 27000.0  448130642.0                  Action|Thriller  ...   
4                   131.0          NaN                      Documentary  ...   

  num_user_for_reviews language  country  content_rating       budget  \
0               3054.0  English      USA           PG-13  237000000.0   
1               1238.0  English      USA           PG-13  300000000.0   
2                994.0  English       UK           PG-13  245000000.0   
3               2701.0  English      USA           PG-13  250000000.0   
4                  NaN      NaN      NaN             NaN          NaN   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   
2      2015.0                  393.0        6.8          2.35   
3      2012.0                23000.0        8.5          2.35   
4         NaN                   12.0        7.1           NaN   

  movie_facebook_likes  
0                33000  
1                    0  
2                85000  
3               164000  
4                    0  

[5 rows x 28 columns]

In [58]:
#Shape of dataset
print("===========Number of rows and coloumns============")
print("\nNumber of rows  : ",data.shape[0])
print("Number of coloumns : ", data.shape[1])

===========Number of rows and coloumns============

Number of rows  :  5043
Number of coloumns :  28


In [21]:
#Info of rows and coloumns
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5043 entries, 0 to 5042
Data columns (total 28 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   color                      5024 non-null   object 
 1   director_name              4939 non-null   object 
 2   num_critic_for_reviews     4993 non-null   float64
 3   duration                   5028 non-null   float64
 4   director_facebook_likes    4939 non-null   float64
 5   actor_3_facebook_likes     5020 non-null   float64
 6   actor_2_name               5030 non-null   object 
 7   actor_1_facebook_likes     5036 non-null   float64
 8   gross                      4159 non-null   float64
 9   genres                     5043 non-null   object 
 10  actor_1_name               5036 non-null   object 
 11  movie_title                5043 non-null   object 
 12  num_voted_users            5043 non-null   int64  
 13  cast_total_facebook_likes  5043 non-null   int64

In [22]:
#Basic summary for numerical columns
data.describe()

num_critic_for_reviews     duration  director_facebook_likes  \
count             4993.000000  5028.000000              4939.000000   
mean               140.194272   107.201074               686.509212   
std                121.601675    25.197441              2813.328607   
min                  1.000000     7.000000                 0.000000   
25%                 50.000000    93.000000                 7.000000   
50%                110.000000   103.000000                49.000000   
75%                195.000000   118.000000               194.500000   
max                813.000000   511.000000             23000.000000   

       actor_3_facebook_likes  actor_1_facebook_likes         gross  \
count             5020.000000             5036.000000  4.159000e+03   
mean               645.009761             6560.047061  4.846841e+07   
std               1665.041728            15020.759120  6.845299e+07   
min                  0.000000                0.000000  1.620000e+02   
25%                133.000000              614.000000  5.340988e+06   
50%                371.500000              988.000000  2.551750e+07   
75%                636.000000            11000.000000  6.230944e+07   
max              23000.000000           640000.000000  7.605058e+08   

       num_voted_users  cast_total_facebook_likes  facenumber_in_poster  \
count     5.043000e+03                5043.000000           5030.000000   
mean      8.366816e+04                9699.063851              1.371173   
std       1.384853e+05               18163.799124              2.013576   
min       5.000000e+00                   0.000000              0.000000   
25%       8.593500e+03                1411.000000              0.000000   
50%       3.435900e+04                3090.000000              1.000000   
75%       9.630900e+04               13756.500000              2.000000   
max       1.689764e+06              656730.000000             43.000000   

       num_user_for_reviews        budget   title_year  \
count           5022.000000  4.551000e+03  4935.000000   
mean             272.770808  3.975262e+07  2002.470517   
std              377.982886  2.061149e+08    12.474599   
min                1.000000  2.180000e+02  1916.000000   
25%               65.000000  6.000000e+06  1999.000000   
50%              156.000000  2.000000e+07  2005.000000   
75%              326.000000  4.500000e+07  2011.000000   
max             5060.000000  1.221550e+10  2016.000000   

       actor_2_facebook_likes   imdb_score  aspect_ratio  movie_facebook_likes  
count             5030.000000  5043.000000   4714.000000           5043.000000  
mean              1651.754473     6.442138      2.220403           7525.964505  
std               4042.438863     1.125116      1.385113          19320.445110  
min                  0.000000     1.600000      1.180000              0.000000  
25%                281.000000     5.800000      1.850000              0.000000  
50%                595.000000     6.600000      2.350000            166.000000  
75%                918.000000     7.200000      2.350000           3000.000000  
max             137000.000000     9.500000     16.000000         349000.000000

In [59]:
#datatypes of each columns
data.dtypes

color                         object
director_name                 object
num_critic_for_reviews       float64
duration                     float64
director_facebook_likes      float64
actor_3_facebook_likes       float64
actor_2_name                  object
actor_1_facebook_likes       float64
gross                        float64
genres                        object
actor_1_name                  object
movie_title                   object
num_voted_users                int64
cast_total_facebook_likes      int64
actor_3_name                  object
facenumber_in_poster         float64
plot_keywords                 object
movie_imdb_link               object
num_user_for_reviews         float64
language                      object
country                       object
content_rating                object
budget                       float64
title_year                   float64
actor_2_facebook_likes       float64
imdb_score                   float64
aspect_ratio                 float64
m

#### Observations

The IMDb movie dataset contains 5,043 movies with 28 attributes, including details on directors, actors, genres, budgets, box office collections, and IMDb scores.
It combines categorical, numerical, and popularity metrics like Facebook likes.
This dataset is ideal for analyzing movie performance, audience sentiment, and industry trends.

## Data Pre-processing

#### Handling Duplicates

In [60]:
# Removing duplicate values

data = data.drop_duplicates()

# Shape after removing duplicates
print("===========Number of rows and coloumns============")
print("\nNumber of rows  : ",data.shape[0])
print("Number of coloumns : ", data.shape[1])


===========Number of rows and coloumns============

Number of rows  :  4998
Number of coloumns :  28


#### Handling Missing Values

In [69]:
# Missing Values
print("Number of missing values for each Column  \n","="*50,"\n",data.isna().sum())

Number of missing values for each Column  
 color                        17
director_name                 0
num_critic_for_reviews        0
duration                      0
director_facebook_likes       0
actor_3_facebook_likes        0
actor_2_name                  0
actor_1_facebook_likes        0
gross                         0
genres                        0
actor_1_name                  0
movie_title                   0
num_voted_users               0
cast_total_facebook_likes     0
actor_3_name                  0
facenumber_in_poster          0
plot_keywords                 0
movie_imdb_link               0
num_user_for_reviews          0
language                      0
country                       0
content_rating                0
budget                        0
title_year                    0
actor_2_facebook_likes        0
imdb_score                    0
aspect_ratio                  0
movie_facebook_likes          0
dtype: int64


In [62]:
# Removing rows with missing values from less critical columns 
data.dropna(subset=["actor_3_facebook_likes"], inplace=True)
data.dropna(subset=["actor_2_facebook_likes"], inplace=True)
data.dropna(subset=["actor_1_facebook_likes"], inplace=True)

In [72]:
# Replacing rows with appropriate values
data["color"].fillna(data["color"].mode()[0] , inplace=True)     # replaced with mode value
data["director_name"].fillna("Unknown", inplace=True)
data["num_critic_for_reviews"].fillna(int(data["num_critic_for_reviews"].mean()) ,inplace=True)     # replaced with mean value 
data["duration"].fillna(data["duration"].mean() , inplace=True)      # replaced with mean value
data["director_facebook_likes"].fillna(0, inplace=True)       #replaced with 0 since Facebook likes naturally fit the idea that "no data" = "no likes"
data["gross"]=data.groupby("genres")["gross"].transform(lambda x:x.fillna(x.mean()))       #Fill missing 'gross' with the mean of its genre group
data["gross"].fillna(data["gross"].mean(), inplace=True)    # Backup plan for gross column 

data["actor_1_name"].fillna("Unknown",inplace=True)
data["actor_2_name"].fillna("Unknown",inplace=True)
data["actor_3_name"].fillna("Unknown",inplace=True)
data["facenumber_in_poster"].fillna(int(data["facenumber_in_poster"].mode()[0]),inplace=True)      # Replaced with mode value
data["plot_keywords"].fillna("Unknown",inplace=True)
data["num_user_for_reviews"].fillna(int(data["num_user_for_reviews"].mean()),inplace=True)
data["language"].fillna("Unknown",inplace=True)
data["country"].fillna("Unknown",inplace=True)
data["content_rating"].fillna("Not Rated",inplace=True)
data["budget"]=data.groupby("genres")["budget"].transform(lambda x: x.fillna(x.mean()))       #Fill missing 'budget' with the mean of its genre group
data["budget"].fillna(data["budget"].mean(), inplace=True)           # Backup plan for budget column 

data["title_year"].fillna(int(data["title_year"].mode()[0]),inplace=True)            # Replaced with mode value
data["aspect_ratio"].fillna(float(data["aspect_ratio"].mode()[0]),inplace=True)          # Replaced with mode value

#### Deriving New coloumns

In [79]:
#Deriving profit column

data["Profit"]=data["gross"]-data["budget"]

#Deriving decade column - to get the decade on film released

data["Decade"]= (data["title_year"]//10)*10

In [83]:
#Deriving Column Successfull/not - cost wise

data["Success_status"] = data.apply(lambda row: "Successful" if row["Profit"] > row["budget"] * 0.1 else "Loss",axis=1)   #considered as successful if profit is greater than 10% of budget

In [89]:
data.sample()

color  director_name  num_critic_for_reviews  duration  \
3076  Color  Oliver Parker                   105.0      97.0   

      director_facebook_likes  actor_3_facebook_likes    actor_2_name  \
3076                     32.0                   327.0  Rupert Everett   

      actor_1_facebook_likes       gross          genres  ... content_rating  \
3076                   893.0  18535191.0  Comedy|Romance  ...          PG-13   

          budget  title_year  actor_2_facebook_likes imdb_score  aspect_ratio  \
3076  14000000.0      1999.0                   692.0        6.9          1.85   

     movie_facebook_likes     Profit  Decade Success_status  
3076                  646  4535191.0  1990.0     Successful  

[1 rows x 31 columns]

#### Filtering

In [96]:
#Recent movies 

recent_movies =data[data["title_year"]>2013]    #for 2013-2016 movies considered as recent
recent_movies

color    director_name  num_critic_for_reviews  duration  \
2                Color       Sam Mendes                   602.0     148.0   
8                Color      Joss Whedon                   635.0     141.0   
10               Color      Zack Snyder                   673.0     183.0   
20               Color    Peter Jackson                   422.0     164.0   
27               Color    Anthony Russo                   516.0     147.0   
...                ...              ...                     ...       ...   
4999             Color      Kirk Loudon                     3.0      95.0   
5012             Color       David Ayer                   233.0     109.0   
5016             Color  Joseph Mazzella                   140.0      90.0   
5019             Color    Marcus Nispel                    43.0      91.0   
5022   Black and White       Jim Chuchu                     6.0      60.0   

      director_facebook_likes  actor_3_facebook_likes        actor_2_name  \
2                         0.0                   161.0        Rory Kinnear   
8                         0.0                 19000.0   Robert Downey Jr.   
10                        0.0                  2000.0        Lauren Cohan   
20                        0.0                   773.0          Adam Brown   
27                       94.0                 11000.0  Scarlett Johansson   
...                       ...                     ...                 ...   
4999                      0.0                   279.0       Johnny Walter   
5012                    453.0                   120.0      Martin Donovan   
5016                      0.0                     9.0        Mikaal Bates   
5019                    158.0                   265.0     Brittany Curran   
5022                      0.0                     4.0       Olwenya Maina   

      actor_1_facebook_likes         gross                       genres  ...  \
2                    11000.0  2.000742e+08    Action|Adventure|Thriller  ...   
8                    26000.0  4.589916e+08      Action|Adventure|Sci-Fi  ...   
10                   15000.0  3.302491e+08      Action|Adventure|Sci-Fi  ...   
20                    5000.0  2.551084e+08            Adventure|Fantasy  ...   
27                   21000.0  4.071973e+08      Action|Adventure|Sci-Fi  ...   
...                      ...           ...                          ...  ...   
4999                   883.0  4.346774e+07        Drama|Sci-Fi|Thriller  ...   
5012                  1000.0  1.049997e+07  Action|Crime|Drama|Thriller  ...   
5016                   313.0  3.266586e+07         Crime|Drama|Thriller  ...   
5019                   630.0  3.884903e+07      Horror|Mystery|Thriller  ...   
5022                   147.0  1.647943e+07                        Drama  ...   

     content_rating        budget  title_year  actor_2_facebook_likes  \
2             PG-13  2.450000e+08      2015.0                   393.0   
8             PG-13  2.500000e+08      2015.0                 21000.0   
10            PG-13  2.500000e+08      2016.0                  4000.0   
20            PG-13  2.500000e+08      2014.0                   972.0   
27            PG-13  2.500000e+08      2016.0                 19000.0   
...             ...           ...         ...                     ...   
4999      Not Rated  7.500000e+04      2014.0                   507.0   
5012              R  3.500000e+07      2014.0                   206.0   
5016      Not Rated  2.500000e+04      2015.0                    25.0   
5019              R  1.643410e+07      2015.0                   512.0   
5022      Not Rated  1.500000e+04      2014.0                    19.0   

     imdb_score  aspect_ratio movie_facebook_likes        Profit  Decade  \
2           6.8          2.35                85000 -4.492582e+07  2010.0   
8           7.5          2.35               118000  2.089916e+08  2010.0   
10          6.9          2.35               197000  8.024906e+07  2010.0   
20          7.5          

In [95]:
#Successfull Movies

Successful=data[data["Success_status"]=="Successful"]
Successful

color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
3     Color  Christopher Nolan                   813.0     164.0   
6     Color          Sam Raimi                   392.0     156.0   
8     Color        Joss Whedon                   635.0     141.0   
9     Color        David Yates                   375.0     153.0   
...     ...                ...                     ...       ...   
5036  Color    Anthony Vallone                   140.0      84.0   
5038  Color        Scott Smith                     1.0      87.0   
5039  Color            Unknown                    43.0      43.0   
5040  Color   Benjamin Roberds                    13.0      76.0   
5042  Color           Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes       actor_2_name  \
0                         0.0                   855.0   Joel David Moore   
3                     22000.0                 23000.0     Christian Bale   
6                         0.0                  4000.0       James Franco   
8                         0.0                 19000.0  Robert Downey Jr.   
9                       282.0                 10000.0   Daniel Radcliffe   
...                       ...                     ...                ...   
5036                      2.0                     2.0     John Considine   
5038                      2.0                   318.0      Daphne Zuniga   
5039                      0.0                   319.0      Valorie Curry   
5040                      0.0                     0.0      Maxwell Moody   
5042                     16.0                    16.0   Brian Herzlinger   

      actor_1_facebook_likes         gross                            genres  \
0                     1000.0  7.605058e+08   Action|Adventure|Fantasy|Sci-Fi   
3                    27000.0  4.481306e+08                   Action|Thriller   
6                    24000.0  3.365303e+08          Action|Adventure|Romance   
8                    26000.0  4.589916e+08           Action|Adventure|Sci-Fi   
9                    25000.0  3.019570e+08  Adventure|Family|Fantasy|Mystery   
...                      ...           ...                               ...   
5036                    45.0  2.281075e+07                       Crime|Drama   
5038                   637.0  2.185859e+07                      Comedy|Drama   
5039                   841.0  3.780378e+07      Crime|Drama|Mystery|Thriller   
5040                     0.0  3.444259e+07             Drama|Horror|Thriller   
5042                    86.0  8.522200e+04                       Documentary   

      ... content_rating        budget  title_year  actor_2_facebook_likes  \
0     ...          PG-13  2.370000e+08      2009.0                   936.0   
3     ...          PG-13  2.500000e+08      2012.0                 23000.0   
6     ...          PG-13  2.580000e+08      2007.0                 11000.0   
8     ...          PG-13  2.500000e+08      2015.0                 21000.0   
9     ...             PG  2.500000e+08      2009.0                 11000.0   
...   ...            ...           ...         ...                     ...   
5036  ...          PG-13  3.250000e+03      2005.0                    44.0   
5038  ...      Not Rated  1.478137e+07      2013.0                   470.0   
5039  ...          TV-14  3.286044e+07      2009.0                   593.0   
5040  ...      Not Rated  1.400000e+03      2013.0                     0.0   
5042  ...             PG  1.100000e+03      2004.0                    23.0   

     imdb_score  aspect_ratio movie_facebook_likes        Profit  Decade  \
0           7.9          1.78                33000  5.235058e+08  2000.0   
3           8.5          2.35               164000  1.981306e+08  2010.0   
6           6.2          2.35                    0  7.853030e+07  2000.0   
8           7.5          2.35               118000  2.089916e+08  2010.0  

In [97]:
#Movies with high IMBD score

High_rated = data[data["imdb_score"]>8]
High_rated

color      director_name  num_critic_for_reviews  duration  \
3     Color  Christopher Nolan                   813.0     164.0   
17    Color        Joss Whedon                   703.0     173.0   
27    Color      Anthony Russo                   516.0     147.0   
43    Color        Lee Unkrich                   453.0     103.0   
58    Color     Andrew Stanton                   421.0      98.0   
...     ...                ...                     ...       ...   
4921  Color       Majid Majidi                    46.0      89.0   
4924  Color          Cary Bell                   140.0      78.0   
4937  Color      Bill Melendez                    43.0      25.0   
4972  Color         Sut Jhally                    16.0      80.0   
5001  Color    Martin Scorsese                    71.0     117.0   

      director_facebook_likes  actor_3_facebook_likes            actor_2_name  \
3                     22000.0                 23000.0          Christian Bale   
17                        0.0                 19000.0       Robert Downey Jr.   
27                       94.0                 11000.0      Scarlett Johansson   
43                      125.0                   721.0       John Ratzenberger   
58                      475.0                   522.0            Fred Willard   
...                       ...                     ...                     ...   
4921                    373.0                    27.0  Amir Farrokh Hashemian   
4924                      0.0                     0.0            Stacie Evans   
4937                     36.0                    27.0           Bill Melendez   
4972                      3.0                     0.0           Seth Ackerman   
5001                  17000.0                   476.0              Levon Helm   

      actor_1_facebook_likes         gross  \
3                    27000.0  4.481306e+08   
17                   26000.0  6.232795e+08   
27                   21000.0  4.071973e+08   
43                   15000.0  4.149845e+08   
58                    1000.0  2.238069e+08   
...                      ...           ...   
4921                    36.0  9.254020e+05   
4924                     0.0  5.746749e+06   
4937                    39.0  1.354330e+08   
4972                   103.0  5.746749e+06   
5001                   725.0  3.219520e+05   

                                         genres  ... content_rating  \
3                               Action|Thriller  ...          PG-13   
17                      Action|Adventure|Sci-Fi  ...          PG-13   
27                      Action|Adventure|Sci-Fi  ...          PG-13   
43    Adventure|Animation|Comedy|Family|Fantasy  ...              G   
58            Adventure|Animation|Family|Sci-Fi  ...              G   
...                                         ...  ...            ...   
4921                               Drama|Family  ...             PG   
4924                                Documentary  ...      Not Rated   
4937                    Animation|Comedy|Family  ...           TV-G   
4972                                Documentary  ...      Not Rated   
5001                          Documentary|Music  ...             PG   

            budget  title_year  actor_2_facebook_likes imdb_score  \
3     2.500000e+08      2012.0                 23000.0        8.5   
17    2.200000e+08      2012.0                 21000.0        8.1   
27    2.500000e+08      2016.0                 19000.0        8.2   
43    2.000000e+08      2010.0                  1000.0        8.3   
58    1.800000e+08      2008.0                   729.0        8.4   
...            ...         ...                     ...        ...   
4921  1.800000e+05      1997.0                    35.0        8.5   
4924  1.800000e+05      2014.0                     0.0        8.7   
4937  1.500000e+05      1965.0                    36.0        8.4   
4972  7.000000e+04      2004.0                     0.0        8.3   
5001  1.557143e+07      1978.0                   57

#### Aggregating 

In [98]:
#Average IMBD score by Genre

avg_score_genres = data.groupby("genres")["imdb_score"].mean()
avg_score_genres

genres
Action                                                             5.972727
Action|Adventure                                                   6.770000
Action|Adventure|Animation|Comedy|Crime|Family|Fantasy             6.200000
Action|Adventure|Animation|Comedy|Drama|Family|Fantasy|Thriller    6.000000
Action|Adventure|Animation|Comedy|Drama|Family|Sci-Fi              7.950000
                                                                     ...   
Sci-Fi|Thriller                                                    6.370000
Thriller                                                           5.326316
Thriller|War                                                       7.900000
Thriller|Western                                                   8.100000
Western                                                            6.583333
Name: imdb_score, Length: 910, dtype: float64

In [100]:
#Total gross collection - country wise

gross_by_country = data.groupby("country")["gross"].sum().sort_values(ascending=False)
gross_by_country

country
USA            1.980400e+11
UK             1.459280e+10
Canada         3.642970e+09
France         3.402603e+09
Germany        2.837204e+09
                   ...     
Kyrgyzstan     2.126511e+06
Afghanistan    1.127331e+06
Finland        6.117090e+05
Philippines    7.007100e+04
Georgia        1.714900e+04
Name: gross, Length: 64, dtype: float64

In [101]:
#Number of movies per Decade

num_of_movies_decade = data.groupby("Decade")["movie_title"].count()
num_of_movies_decade

Decade
1910.0       1
1920.0       5
1930.0      15
1940.0      24
1950.0      28
1960.0      72
1970.0     111
1980.0     287
1990.0     782
2000.0    2177
2010.0    1473
Name: movie_title, dtype: int64

#### Observations

The dataset was cleaned by handling duplicates and missing values using mean, mode, and group-wise replacement. New columns like Profit, Decade, and Success Status gave insights into financial trends and movie success. Filtering and aggregation further revealed patterns in genres, countries, and decades, highlighting profitability and audience preferences.

## Exploratory Data Analysis (EDA)